subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/JavaCompilerArgumentsBuilder.java

```diff
@@ -181,17 +186,32 @@ Due to a bug (https://builds.gradle.org/viewLog.html?buildId=284033&tab=buildRes
         args.add(USE_UNSHARED_COMPILER_TABLE_OPTION);
     }
 
-    private void addUserProvidedArgs() {
+    private void addUserProvidedArgs(List<String> compilerArgs) {
         if (!includeMainOptions) {
             return;
         }
-
-        List<String> compilerArgs = spec.getCompileOptions().getCompilerArgs();
         if (compilerArgs != null) {
             args.addAll(compilerArgs);
         }
     }
 
+    private String extractSourcepathFrom(List<String> compilerArgs) {
+        Iterator<String> argIterator = compilerArgs.iterator();
+        String userProvidedSourcepath = "";
+        while (argIterator.hasNext()) {
+            String current = argIterator.next();
+            if (current.equals("-sourcepath") || current.equals("--source-path")) {
+                DeprecationLogger.nagUserOfDeprecated(
+                    "Specifying the source path in the CompilerOptions compilerArgs property",
+                    "Instead, use the CompilerOptions sourcepath property directly");
+                argIterator.remove();
+                userProvidedSourcepath = argIterator.next();
```

Comment:
:o: This may throw a `NoSuchElementException`, but I'm not sure we care: this would be a user error.